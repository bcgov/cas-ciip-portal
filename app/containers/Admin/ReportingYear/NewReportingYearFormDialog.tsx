import React from 'react';
import moment from 'moment-timezone';
import {Modal, Container, Button} from 'react-bootstrap';
import globalFormStyles from '../../Forms/FormSharedStyles';
import JsonSchemaForm from 'react-jsonschema-form';
import {JSONSchema6} from 'json-schema';
import FormObjectFieldTemplate from 'containers/Forms/FormObjectFieldTemplate';
import FormFieldTemplate from 'containers/Forms/FormFieldTemplate';
import newReportingYearSchema from './create_reporting_year.json';
import AltDateInput from 'components/Forms/AltDateInput';
import AltDateTimeInput from 'components/Forms/AltDateTimeInput';
import {validateAllDates, validateUniqueKey} from './reportingYearValidation';

const TIME_ZONE = 'America/Vancouver';

function transformUiSchema(json) {
  const now = moment.tz(TIME_ZONE);
  const thisYear = Number(now.year());

  Object.keys(json).forEach((field) => {
    if (field.includes('reportingPeriod')) {
      json[field]['ui:options'].yearsRange = [thisYear - 2, thisYear + 10];
    } else if (field.includes('application')) {
      json[field]['ui:options'].yearsRange = [thisYear, thisYear + 10];
    }
  });

  return json;
}

interface Props {
  show: boolean;
  clearForm: () => void;
  createReportingYear: ({formData}) => void;
  existingYearKeys: number[];
}

const NewReportingYearFormDialog: React.FunctionComponent<Props> = ({
  show,
  clearForm,
  createReportingYear,
  existingYearKeys
}) => {
  const uiSchema = transformUiSchema(newReportingYearSchema.uiSchema);

  const handleSubmit = (e) => {
    if (e.errors.length === 0) {
      createReportingYear(e);
    }
  };

  return (
    <>
      <Modal centered size="xl" show={show} onHide={clearForm}>
        <Modal.Header closeButton>
          <Modal.Title>New Reporting Year</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Container>
            <JsonSchemaForm
              omitExtraData
              liveOmit
              noHtml5Validate
              schema={newReportingYearSchema.schema as JSONSchema6}
              uiSchema={uiSchema}
              formData={{}}
              FieldTemplate={FormFieldTemplate}
              ObjectFieldTemplate={FormObjectFieldTemplate}
              widgets={{AltDateInput, AltDateTimeInput}}
              showErrorList={false}
              validate={(formData, errors) => {
                validateUniqueKey(existingYearKeys, formData, errors);
                return validateAllDates(
                  null,
                  formData,
                  errors,
                  newReportingYearSchema.uiSchema
                );
              }}
              onSubmit={handleSubmit}
            >
              <Button type="submit" variant="primary">
                Save Reporting Year
              </Button>
            </JsonSchemaForm>
          </Container>
        </Modal.Body>
      </Modal>
      <style jsx global>
        {globalFormStyles}
      </style>
    </>
  );
};

export default NewReportingYearFormDialog;
