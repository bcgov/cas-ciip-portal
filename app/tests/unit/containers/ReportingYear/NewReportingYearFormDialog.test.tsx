import React from "react";
import { shallow } from "enzyme";
import NewReportingYearFormDialog from "containers/Admin/ReportingYear/NewReportingYearFormDialog";

describe("NewReportingYearFormDialog", () => {
  it("Should match last accepted snapshot", async () => {
    const r = shallow(
      <NewReportingYearFormDialog
        show
        clearForm={jest.fn()}
        createReportingYear={jest.fn()}
        existingYearKeys={[2019, 2020]}
      />
    );
    expect(r).toMatchSnapshot();
  });
});
