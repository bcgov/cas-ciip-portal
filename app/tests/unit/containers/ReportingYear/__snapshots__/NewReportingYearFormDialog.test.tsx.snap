// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewReportingYearFormDialog Should match last accepted snapshot 1`] = `
<Fragment>
  <Bootstrap(Modal)
    centered={true}
    onHide={[MockFunction]}
    show={true}
    size="xl"
  >
    <ModalHeader
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle>
        New Reporting Year
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <Container
        fluid={false}
      >
        <Form
          ErrorList={[Function]}
          FieldTemplate={[Function]}
          ObjectFieldTemplate={[Function]}
          disabled={false}
          formData={Object {}}
          liveOmit={true}
          liveValidate={false}
          noHtml5Validate={true}
          noValidate={false}
          omitExtraData={true}
          onSubmit={[Function]}
          safeRenderCompletion={false}
          schema={
            Object {
              "properties": Object {
                "applicationCloseTime": Object {
                  "title": "Application Close Time",
                  "type": "string",
                },
                "applicationOpenTime": Object {
                  "title": "Application Open Time",
                  "type": "string",
                },
                "reportingPeriodEnd": Object {
                  "title": "Reporting Period End",
                  "type": "string",
                },
                "reportingPeriodStart": Object {
                  "title": "Reporting Period Start",
                  "type": "string",
                },
                "reportingYear": Object {
                  "title": "Reporting Year",
                  "type": "integer",
                },
                "swrsDeadline": Object {
                  "title": "Industrial GHG Reporting (SWRS) Deadline",
                  "type": "string",
                },
              },
              "required": Array [
                "reportingYear",
                "swrsDeadline",
                "reportingPeriodStart",
                "reportingPeriodEnd",
                "applicationOpenTime",
                "applicationCloseTime",
              ],
              "type": "object",
            }
          }
          showErrorList={false}
          uiSchema={
            Object {
              "applicationCloseTime": Object {
                "ui:options": Object {
                  "hideClearButton": true,
                  "hideNowButton": true,
                  "yearsRange": Array [
                    2020,
                    2030,
                  ],
                },
                "ui:widget": "AltDateTimeInput",
              },
              "applicationOpenTime": Object {
                "ui:options": Object {
                  "hideClearButton": true,
                  "hideNowButton": true,
                  "yearsRange": Array [
                    2020,
                    2030,
                  ],
                },
                "ui:widget": "AltDateTimeInput",
              },
              "reportingPeriodEnd": Object {
                "ui:options": Object {
                  "hideClearButton": true,
                  "hideNowButton": true,
                  "yearsRange": Array [
                    2018,
                    2030,
                  ],
                },
                "ui:widget": "AltDateInput",
              },
              "reportingPeriodStart": Object {
                "ui:options": Object {
                  "hideClearButton": true,
                  "hideNowButton": true,
                  "yearsRange": Array [
                    2018,
                    2030,
                  ],
                },
                "ui:widget": "AltDateInput",
              },
              "swrsDeadline": Object {
                "ui:options": Object {
                  "hideClearButton": true,
                  "hideNowButton": true,
                  "yearsRange": Array [
                    2020,
                    2030,
                  ],
                },
                "ui:widget": "AltDateInput",
              },
            }
          }
          validate={[Function]}
          widgets={
            Object {
              "AltDateInput": [Function],
              "AltDateTimeInput": [Function],
            }
          }
        >
          <Button
            active={false}
            disabled={false}
            type="submit"
            variant="primary"
          >
            Save Reporting Year
          </Button>
        </Form>
      </Container>
    </ModalBody>
  </Bootstrap(Modal)>
  <JSXStyle
    id="2362864946"
  >
    .
    r
    j
    s
    f
     
    .
    f
    o
    r
    m
    -
    r
    o
    w
    {
    m
    a
    r
    g
    i
    n
    :
    2
    0
    p
    x
     
    0
     
    2
    0
    p
    x
    ;
    }
    .
    e
    m
    i
    s
    s
    i
    o
    n
    -
    f
    o
    r
    m
     
    .
    r
    j
    s
    f
     
    .
    f
    o
    r
    m
    -
    r
    o
    w
    {
    m
    a
    r
    g
    i
    n
    :
    0
    ;
    }
    .
    n
    o
    -
    c
    o
    l
    -
    p
    a
    d
    d
    i
    n
    g
     
    .
    f
    o
    r
    m
    -
    r
    o
    w
    &gt;
    .
    c
    o
    l
    ,
    .
    n
    o
    -
    c
    o
    l
    -
    p
    a
    d
    d
    i
    n
    g
     
    .
    f
    o
    r
    m
    -
    r
    o
    w
    &gt;
    [
    c
    l
    a
    s
    s
    *
    =
    '
    c
    o
    l
    -
    '
    ]
    {
    p
    a
    d
    d
    i
    n
    g
    :
    0
    ;
    }
    .
    f
    o
    r
    m
    -
    s
    u
    b
    m
    i
    t
    {
    b
    o
    r
    d
    e
    r
    :
    1
    p
    x
     
    s
    o
    l
    i
    d
     
    #
    b
    b
    b
    ;
    p
    a
    d
    d
    i
    n
    g
    :
    3
    0
    p
    x
    ;
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    :
    #
    e
    0
    e
    0
    e
    0
    ;
    b
    o
    r
    d
    e
    r
    -
    r
    a
    d
    i
    u
    s
    :
    6
    p
    x
    ;
    }
    .
    e
    r
    r
    o
    r
    s
    {
    m
    a
    r
    g
    i
    n
    -
    l
    e
    f
    t
    :
    2
    0
    p
    x
    ;
    p
    a
    d
    d
    i
    n
    g
    :
    2
    0
    p
    x
    ;
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    :
    #
    c
    e
    5
    c
    5
    c
    ;
    c
    o
    l
    o
    r
    :
    w
    h
    i
    t
    e
    ;
    f
    o
    n
    t
    -
    s
    i
    z
    e
    :
    2
    0
    p
    x
    ;
    }
    u
    l
    .
    e
    r
    r
    o
    r
    -
    d
    e
    t
    a
    i
    l
    {
    p
    a
    d
    d
    i
    n
    g
    :
    0
    ;
    l
    i
    s
    t
    -
    s
    t
    y
    l
    e
    :
    n
    o
    n
    e
    ;
    }
    u
    l
    .
    e
    r
    r
    o
    r
    -
    d
    e
    t
    a
    i
    l
     
    l
    i
    .
    t
    e
    x
    t
    -
    d
    a
    n
    g
    e
    r
    {
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    :
    #
    c
    e
    5
    c
    5
    c
    ;
    c
    o
    l
    o
    r
    :
    w
    h
    i
    t
    e
     
    !
    i
    m
    p
    o
    r
    t
    a
    n
    t
    ;
    p
    a
    d
    d
    i
    n
    g
    :
    5
    p
    x
    ;
    }
    .
    e
    x
    i
    t
    -
    b
    u
    t
    t
    o
    n
    {
    b
    a
    c
    k
    g
    r
    o
    u
    n
    d
    :
    w
    h
    i
    t
    e
     
    !
    i
    m
    p
    o
    r
    t
    a
    n
    t
    ;
    c
    o
    l
    o
    r
    :
    #
    0
    0
    3
    3
    6
    6
    ;
    m
    a
    r
    g
    i
    n
    -
    l
    e
    f
    t
    :
    5
    p
    x
    ;
    }
  </JSXStyle>
</Fragment>
`;
